client/my-sites/stats/features/modules/stats-locations/stats-locations.tsx

```diff
@@ -62,29 +66,64 @@ type SelectOptionType = {
 
 interface StatsModuleLocationsProps extends StatsDefaultModuleProps {
 	initialGeoMode?: string;
+	query: StatsQueryType;
+	context: {
+		query: {
+			[ key: string ]: string | undefined;
+		};
+	};
 }
 
 const StatsLocations: React.FC< StatsModuleLocationsProps > = ( {
 	initialGeoMode,
 	period,
 	query,
 	summaryUrl,
+	context,
+	summary,
 } ) => {
 	const translate = useTranslate();
 	const siteId = useSelector( getSelectedSiteId ) as number;
+	const siteSlug = useSelector( ( state ) => getSiteSlug( state, siteId ) );
 	const statType = STAT_TYPE_COUNTRY_VIEWS;
 	const isOdysseyStats = config.isEnabled( 'is_running_in_jetpack_site' );
 	const supportUrl = isOdysseyStats
 		? `${ JETPACK_SUPPORT_URL_TRAFFIC }#views-by-locations`
 		: LOCATIONS_SUPPORT_URL;
+	const dispatch = useDispatch();
+	const [ selectedOption, setSelectedOption ] = useState( () => {
+		const urlGeoMode = initialGeoMode || context.query.geoMode;
+		return urlGeoMode && urlGeoMode in GEO_MODES ? urlGeoMode : OPTION_KEYS.COUNTRIES;
+	} );
+	const isStatsNavigationImprovementEnabled = config.isEnabled( 'stats/navigation-improvement' );
+
+	const [ countryFilter, setCountryFilter ] = useState< string | null >( null );
 
-	const urlGeoMode =
-		initialGeoMode || new URLSearchParams( window.location.search ).get( 'geoMode' );
-	const [ selectedOption, setSelectedOption ] = useState(
-		urlGeoMode && urlGeoMode in GEO_MODES ? urlGeoMode : OPTION_KEYS.COUNTRIES
+	const basePath = useMemo(
+		() => `/stats/${ period.period }/locations/${ siteSlug }`,
+		[ period.period, siteSlug ]
 	);
 
-	const [ countryFilter, setCountryFilter ] = useState< string | null >( null );
+	useEffect( () => {
+		if ( ! summary ) {
+			return;
+		}
+
+		const geoModeParam = context.query[ 'geoMode' ];
+		const isValidGeoModeParam =
+			geoModeParam && Object.values( OPTION_KEYS ).includes( geoModeParam );
+
+		if ( ! isValidGeoModeParam ) {
+			return;
+		}
+
+		// If URL has valid param and it's different from state, update state
+		if ( geoModeParam !== selectedOption ) {
+			const updatedQuery = { ...context.query, [ 'geoMode' ]: selectedOption };
+			const queryString = new URLSearchParams( updatedQuery ).toString();
+			page( `${ basePath }?${ queryString }` );
```

Comment:
Keep the `geoMode` query param in the permalink up-to-date as this is now used as filename for the CSV download

Reply:
I have reverted all changes related to geoMode in favor of https://github.com/Automattic/wp-calypso/pull/103229